easybuild/easyblocks/x/xmipp.py

```diff
@@ -0,0 +1,116 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with EasyBuild.  If not, see <http://www.gnu.org/licenses/>.
+##
+"""
+EasyBuild support for building and installing Xmipp, implemented as an easyblock
+
+@author: Jens Timmerman (Ghent University)
+"""
+import glob
+import os
+
+from easybuild.framework.easyblock import EasyBlock
+from easybuild.tools.filetools import mkdir, extract_file
+from easybuild.tools.modules import get_software_root, get_software_version
+from easybuild.tools.run import run_cmd
+
+
+class EB_Xmipp(EasyBlock):
+    """Support for building/installing Xmipp."""
+
+    def __init__(self, *args, **kwargs):
+        """Easyblock constructor, enable building in installation directory."""
+        super(self.__class__, self).__init__(*args, **kwargs)
+        self.build_in_installdir = True
+
+    def extract_step(self):
+        """Extract sources."""
+        # strip off 'xmipp' part to avoid having everything in a 'xmipp' subdirectory
+        self.cfg['unpack_options'] = "--strip-components=1"
+        super(self.__class__, self).extract_step()
+
+    def configure_step(self):
+        """Configure xmipp build via a provided wrapper around sconss."""
+        # install step expects these to exist
+        mkdir('lib')
+        mkdir('bin')
+        # run interactive 'config' script to generate install.info file
+        try:
+            # ln -s `which python` xmipp_python
+            orig = os.path.join(get_software_root('Python'), 'bin', 'python')
+            dest = 'xmipp_python'
+            os.symlink(orig, dest)
+            # ln -s $EBROOTPYTHON/lib/python2.7 lib/python2.7
+            self.python_short_ver = ".".join(get_software_version('Python').split(".")[0:2])
+            orig = os.path.join(get_software_root('Python'), 'lib', 'python%s' % self.python_short_ver)
+            dest = os.path.join('lib', 'python%s' % self.python_short_ver)
+            os.symlink(orig, dest)
+        except OSError as err:
+            self.log.error("Failed to symlink %s to %s: %s" % (orig, dest, err))
+
+        # extract some dependencies that we really need and can't find anywhere else
+        # alglib tarball has version in name, so lets find it with a glob
+        alglib_tar = glob.glob(os.path.join(os.getcwd(), 'external', 'alglib*.tgz'))[0]
+        for src in ['bilib.tgz', 'bilib.tgz', 'condor.tgz', alglib_tar]:
+            external_path = os.path.join(os.getcwd(), 'external')
+            extract_file(os.path.join(external_path, src), external_path)
+
+        cmd = ('{preconfigopts} PATH=$PWD:$PATH xmipp_python external/scons/scons.py mode=configure -j {parallel}'
+               ' --config=force  profile=no fast=yes warn=no release=yes gtest=no cuda=no debug=no matlab=no'
+               ' java=no LINKERFORPROGRAMS="$CXX" MPI_BINDIR="$EBROOTIMPI" JAVA_HOME="$JAVA_HOME" JAVAC=javac CC="$CC"'
```

Comment:
only works if `impi` is loaded, so add a check to make sure it is?

isn't this supposed to point to the `bin` subdir btw?

or just use something like `os.path.join(get_software_root(self.toolchain.MPI_MODULE_NAME[0]), 'bin')`